### ISSUE TITLE
First spawn of a ParticleEffectBundle with a "once" spawner does not emit particles

### ISSUE THREAD

**Issue by Jaso333** [djeedai/bevy_hanabi#319]
**Crate versions**
`bevy` version: 0.13.2
`bevy_hanabi` version: 0.10.0

**Describe the bug**
When spawning an entity with a particle effect that is configured with `Spawner::once` and `starts_immediately` set to `true`, it does not emit any particles. However, it does happen the second time an entity of the exact same configuration is spawned.

**Expected behavior**
The first time the entity is spawned with the above configuration, particles should emit.

**To Reproduce**
1. Insert a resource containing an effect asset that uses a "once" spawner with `starts_immediately` set to `true`. This happens in a `Startup` system.
2. In an `Update` system that runs if an event is received, a `ParticleEffectBundle` is spawned in a new entity, which references the effect asset created earlier.
3. Trigger the event once, no particles are emitted from the spawned entity.
4. Trigger the event a second time and the particles are correctly emited from the second spawned entity.

An minimal reproducable example of this happening can be shown here: https://github.com/Jaso333/hanabi-issue

**Screenshots**
N/A

**Comment by djeedai** [comment]
Thanks for the repro, I will have a look.

**Comment by djeedai** [comment]
I just tested and this works on `main` with latest changes. There were a bunch of bug fixes recently so I assume this was fixed among them. Closing for now, please feel free to re-open if there's still an issue. Thanks!

**Comment by vhakr** [comment]
I'm still having this issue, I tried the master branch and tried copying over example configurations. Strangely the spawner.rs example works perfectly. I've also noticed the the first activation of burst doesn't work.

**Comment by PraxTube** [comment]
I am still seeing this behaviour on 0.14 (with bevy 0.15.1). I have the same setup (a resource with a handle to an EffectAsset and spawning the effect multiple times at runtime). However the first effect is never visible.

I can reproduce the issue with the `firework` example on `main`. This is how it looks. The first burst is not right. If I were to use `once` instead of `burst` then I would only see the buggy one.

https://github.com/user-attachments/assets/f896882e-56b0-49ba-a2b3-a4e8e18183bc

In my game it seems like the first effect isn't spawning at all, however that is not the case (and I am assuming the same is true for the original bug report). The effect is indeed spawned, however the first burst is just buggy and not really visible, which leads to the conclusion it's not spawned (I changed the `once` to a `burst` and you can see that the first burst of the first effect is "missing").


https://github.com/user-attachments/assets/33032ac0-f41a-4feb-b681-ac495e3a27f0

**Comment by kim-fg** [comment]
I tried testing and working on this issue a bit since I'm experiencing the same issue. All my testing so far has been on the Firework example. 
Spawner::once with spawn_immediately being true does not seem to spawn at all. Spawner::burst does spawn immediately, but the first one does not apply initial velocities (picture below). Gravity does seem to be applied correctly for the first burst though, so my assumption is that it's a "first frame" issue with both spawner types.

![Firework example first burst](https://github.com/user-attachments/assets/3ad27865-1072-48c6-836e-553609b99b23)

I haven't gotten much closer to figuring out the root cause, but I'm fairly certain it is not an issue with the Spawner or EffectSpawner structs. I'm moving on to the `spawn::tick_initializers` system but I don't think I'll make much progress there either.

**Comment by djeedai** [comment]
@kim-fg have you taken into account the fact that the very first frame of any Bevy app has a delta time of zero, which breaks everything? This is a known issue, we should avoid testing that first frame. Hanabi needs to fall back to a non-zero timestep but currently doesn't. I don't think it's related to the current issue but any testing should be careful to avoid that edge case.

**Comment by PraxTube** [comment]
The time delta thing is a good point, though it's definitely not the cause for this issue (or at least not the only cause). The glitch appears whenever you spawn the effect for the first time, regardless of when that first spawn is.

**Comment by djeedai** [comment]
@kim-fg yes agreed, just wanted to make sure we don't mix issues. Did you repro on latest version? And on `main`? Because the original issue is from Hanabi 0.10 which is a bit old (the code changed quite a bit)

**Comment by PraxTube** [comment]
No clue what which version @kim-fg tested, though I just pulled on main and the issue is still present, I also tested to see if the delta thing is the actual issue and it is somewhat an issue, the weird physics behaviour seems to go away (mostly) when I use `setup.run_if(once_after_delay(Duration::from_secs_f32(1.0)))`, however I still get unexpected behaviour (it's a little unfortunate but it's much easier to see than it is to explain, so here is another video).

https://github.com/user-attachments/assets/7f3cf71a-ef7f-4195-a08f-b3d1e6e3970d

You can see that the first burst is for some reason smaller than the rest of them. There definitely seems to be something fishy about the first burst (and also the `once`, which looks pretty much the same).

And again, spawning the same effect a second time after that works as expected (whether burst or once).

**Comment by djeedai** [comment]
On `main` if you know how to use RenderDoc you can configure `DebugSettings` to capture N frames each time an effect spawns. That will show you the data on GPU, could be a good help to debug.

**Comment by djeedai** [comment]
...and thanks for confirming it reproduces on `main` @PraxTube that's quite valuable.

**Comment by PraxTube** [comment]
> This makes me think that the initial velocity is not applied as an impulse but rather as an accumulation lol. Either way this is a good find as it means the intital velocity for the first burst is actually applied. Did you try this for Spawner::once as well?

Oh yeah it kinda seems like it's accumulating and not setting, that would be pretty weird. Yeah I tested the once as well, it looked like the first burst of `Spawner::burst`.

**Comment by kim-fg** [comment]
So I simply printed the init_shader_code string and got something to look at. Specifically interesting for this is the initialization in the main fn (same code used for clone):

```wgsl
// main
// ...

// Initialize new particle    
var particle = Particle();    
set_position_sphere_92C0812A7476F536(&particle);
let var0 = frand();
particle.velocity = normalize(particle.position - (vec3<f32>(0.,0.,0.))) * (((var0) * (20.)) + (60.));
particle.age = rand_uniform_f(0., 0.2);
particle.lifetime = rand_normal_f(0.8, 1.2);

// ...
```

Although highly unlikely - if the `set_position_sphere_...` function sets the particle position to `vec3(0.,0.,0.)` then the velocity evaluates to `vec3(0., 0., 0.)` as well which does lead to the issues we've been seeing.
I was also thinking that `frand()` could return `0.` to `var0`, but since we add `.60` this wouldn't lead to the issues.

So okay, how could `set_position_sphere` put the particle at `(0., 0., 0.)`?

```wgsl
// set_position_sphere_...

// Sphere center    
let c = vec3<f32>(0.,0.,0.);
// Sphere radius    
let r = pow(frand(), 1./3.) * (0.1);    
// Spawn randomly along the sphere surface using Archimedes's theorem    
let theta = frand() * tau;    
let z = frand() * 2. - 1.;    
let phi = acos(z);    
let sinphi = sin(phi);    
let x = sinphi * cos(theta);    
let y = sinphi * sin(theta);    
let dir = vec3<f32>(x, y, z);    
(*particle).position = c + r * dir;
```

There's that `frand()` again..
If it returns `0.` when the radius is set - `pow(frand(), 1./3.) * (0.1)` evals to `0.` which would put all particles at `(0., 0., 0.)` IF it's consistent for the entire particle group.

This could all be the wrong path, but my naive theory before digging deeper is that whatever `frand()` uses to generate numbers is not initialized during the first burst. If I'm wrong we can at least _remove this from the equation_ and check other things.

I've added the whole "prettyfied" printed code as a txt file below as well. It doesn't look quite right in places but hey I did it manually so :shrug:

[pretty_printed_shader.txt](https://github.com/user-attachments/files/18588933/pretty_printed_shader.txt)

**Comment by barsoosayque** [comment]
It seems that https://github.com/djeedai/bevy_hanabi/commit/780d0bc30a223e949591f7153fa25e079406602a fixed the issue -- on `bevy_hanabi = "0.17"` I don't even have to pre-spawn such oneshot effects anymore, they just work.
